### ISSUE TITLE
Link for public polls incorrect

### ISSUE THREAD

**Issue by scroom** [nextcloud/polls#86]
Today I've tried the first time to create a public poll with the polls app and discovered a bug. The link which is created to share it with others is incorrect. In my case (tested on two servers) the created link looked like this

<code>https://example.com/owncloud//owncloud/index.php/apps/polls/poll/79587hdkhd3zrh83</code>

it should look like this:

<code>https://example.com/owncloud/index.php/apps/polls/poll/79587hdkhd3zrh83</code>

I can access the poll if I’m changing the link manually.

**Comment by v1r0x** [comment]
This is weird, it should has been fixed with #64 
But I'll have a look at it.

**Comment by v1r0x** [comment]
I could not reproduce it. Link looks good (`https://domain.tld/index.php/apps/polls/poll/....`).
Which version of oc/nc do you use? And what branch/version of polls do you use?

**Comment by scroom** [comment]
I'm using Nextcloud 10.0.1 (stable) and the enh-group-display branch

**Comment by v1r0x** [comment]
Could you try the master branch and if it still doesn't work the latest release?

**Comment by scroom** [comment]
Master needed minutes to load when I wanted to open a new poll (maybe you haven't merged here the changes made in the enh-group-display branch?) and then produced an internal server error when I wanted to finish the opening of the new poll. Here comes the log:

```
{"reqId":"AXcXoRuNzUQ0QZCaP2Ku","remoteAddr":"78.94.103.130","app":"index","message":"Exception: {\"Exception\":\"Doctrine\\\\DBAL\\\\Exception\\\\InvalidFieldNameException\",\"Message\":\"An exception occurred while executing 'INSERT INTO `oc_polls_events`(`title`,`description`,`owner`,`created`,`hash`,`is_anonymous`,`access`,`type`) VALUES(?,?,?,?,?,?,?,?)' with params [\\\"Open\\\", \\\"Test\\\", \\\"user\\\", \\\"2016-10-13 07:44:42\\\", \\\"mxfGcVKVup0M7x8J\\\", \\\"0\\\", \\\"public\\\", 0]:\\n\\nSQLSTATE[42S22]: Column not found: 1054 Unknown column 'is_anonymous' in 'field list'\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/html\\\/cloud\\\/3rdparty\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/DBALException.php(116): Doctrine\\\\DBAL\\\\Driver\\\\AbstractMySQLDriver->convertException('An exception oc...', Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOException))\\n#1 \\\/var\\\/www\\\/html\\\/cloud\\\/3rdparty\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/Statement.php(175): Doctrine\\\\DBAL\\\\DBALException::driverExceptionDuringQuery(Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOMySql\\\\Driver), Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOException), 'INSERT INTO `oc...', Array)\\n#2 \\\/var\\\/www\\\/html\\\/cloud\\\/lib\\\/private\\\/legacy\\\/db\\\/statementwrapper.php(76): Doctrine\\\\DBAL\\\\Statement->execute()\\n#3 \\\/var\\\/www\\\/html\\\/cloud\\\/lib\\\/public\\\/AppFramework\\\/Db\\\/Mapper.php(252): OC_DB_StatementWrapper->execute()\\n#4 \\\/var\\\/www\\\/html\\\/cloud\\\/lib\\\/public\\\/AppFramework\\\/Db\\\/Mapper.php(126): OCP\\\\AppFramework\\\\Db\\\\Mapper->execute('INSERT INTO `*P...', Array)\\n#5 \\\/var\\\/www\\\/html\\\/cloud\\\/apps\\\/polls\\\/controller\\\/pagecontroller.php(334): OCP\\\\AppFramework\\\\Db\\\\Mapper->insert(Object(OCA\\\\Polls\\\\Db\\\\Event))\\n#6 [internal function]: OCA\\\\Polls\\\\Controller\\\\PageController->insertPoll('event', 'Open', 'Test', 'user', '[1476431100,147...', '', 'public', '', NULL)\\n#7 \\\/var\\\/www\\\/html\\\/cloud\\\/lib\\\/private\\\/AppFramework\\\/Http\\\/Dispatcher.php(160): call_user_func_array(Array, Array)\\n#8 \\\/var\\\/www\\\/html\\\/cloud\\\/lib\\\/private\\\/AppFramework\\\/Http\\\/Dispatcher.php(90): OC\\\\AppFramework\\\\Http\\\\Dispatcher->executeController(Object(OCA\\\\Polls\\\\Controller\\\\PageController), 'insertPoll')\\n#9 \\\/var\\\/www\\\/html\\\/cloud\\\/lib\\\/private\\\/AppFramework\\\/App.php(111): OC\\\\AppFramework\\\\Http\\\\Dispatcher->dispatch(Object(OCA\\\\Polls\\\\Controller\\\\PageController), 'insertPoll')\\n#10 \\\/var\\\/www\\\/html\\\/cloud\\\/lib\\\/private\\\/AppFramework\\\/Routing\\\/RouteActionHandler.php(47): OC\\\\AppFramework\\\\App::main('PageController', 'insertPoll', Object(OC\\\\AppFramework\\\\DependencyInjection\\\\DIContainer), Array)\\n#11 [internal function]: OC\\\\AppFramework\\\\Routing\\\\RouteActionHandler->__invoke(Array)\\n#12 \\\/var\\\/www\\\/html\\\/cloud\\\/lib\\\/private\\\/Route\\\/Router.php(293): call_user_func(Object(OC\\\\AppFramework\\\\Routing\\\\RouteActionHandler), Array)\\n#13 \\\/var\\\/www\\\/html\\\/cloud\\\/lib\\\/base.php(983): OC\\\\Route\\\\Router->match('\\\/apps\\\/polls\\\/ins...')\\n#14 \\\/var\\\/www\\\/html\\\/cloud\\\/index.php(48): OC::handleRequest()\\n#15 {main}\",\"File\":\"\\\/var\\\/www\\\/html\\\/cloud\\\/3rdparty\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/Driver\\\/AbstractMySQLDriver.php\",\"Line\":71}","level":3,"time":"2016-10-13T07:44:42+00:00","method":"POST","url":"\/cloud\/index.php\/apps\/polls\/insert","user":"user"}
{"reqId":"hhWCqzUg18JWgIg4lC\/z","remoteAddr":"78.94.103.130","app":"index","message":"Exception: {\"Exception\":\"Doctrine\\\\DBAL\\\\Exception\\\\InvalidFieldNameException\",\"Message\":\"An exception occurred while executing 'INSERT INTO `oc_polls_events`(`title`,`description`,`owner`,`created`,`hash`,`is_anonymous`,`access`,`expire`,`type`) VALUES(?,?,?,?,?,?,?,?,?)' with params [\\\"Open\\\", \\\"Test\\\", \\\"user\\\", \\\"2016-10-13 07:45:08\\\", \\\"e8SR7FPmgPLDzng6\\\", \\\"0\\\", \\\"public\\\", \\\"2016-10-12 22:00:00\\\", 0]:\\n\\nSQLSTATE[42S22]: Column not found: 1054 Unknown column 'is_anonymous' in 'field list'\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/html\\\/cloud\\\/3rdparty\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/DBALException.php(116): Doctrine\\\\DBAL\\\\Driver\\\\AbstractMySQLDriver->convertException('An exception oc...', Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOException))\\n#1 \\\/var\\\/www\\\/html\\\/cloud\\\/3rdparty\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/Statement.php(175): Doctrine\\\\DBAL\\\\DBALException::driverExceptionDuringQuery(Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOMySql\\\\Driver), Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOException), 'INSERT INTO `oc...', Array)\\n#2 \\\/var\\\/www\\\/html\\\/cloud\\\/lib\\\/private\\\/legacy\\\/db\\\/statementwrapper.php(76): Doctrine\\\\DBAL\\\\Statement->execute()\\n#3 \\\/var\\\/www\\\/html\\\/cloud\\\/lib\\\/public\\\/AppFramework\\\/Db\\\/Mapper.php(252): OC_DB_StatementWrapper->execute()\\n#4 \\\/var\\\/www\\\/html\\\/cloud\\\/lib\\\/public\\\/AppFramework\\\/Db\\\/Mapper.php(126): OCP\\\\AppFramework\\\\Db\\\\Mapper->execute('INSERT INTO `*P...', Array)\\n#5 \\\/var\\\/www\\\/html\\\/cloud\\\/apps\\\/polls\\\/controller\\\/pagecontroller.php(334): OCP\\\\AppFramework\\\\Db\\\\Mapper->insert(Object(OCA\\\\Polls\\\\Db\\\\Event))\\n#6 [internal function]: OCA\\\\Polls\\\\Controller\\\\PageController->insertPoll('event', 'Open', 'Test', 'user', '[1476431100,147...', '1476223200', 'public', '', NULL)\\n#7 \\\/var\\\/www\\\/html\\\/cloud\\\/lib\\\/private\\\/AppFramework\\\/Http\\\/Dispatcher.php(160): call_user_func_array(Array, Array)\\n#8 \\\/var\\\/www\\\/html\\\/cloud\\\/lib\\\/private\\\/AppFramework\\\/Http\\\/Dispatcher.php(90): OC\\\\AppFramework\\\\Http\\\\Dispatcher->executeController(Object(OCA\\\\Polls\\\\Controller\\\\PageController), 'insertPoll')\\n#9 \\\/var\\\/www\\\/html\\\/cloud\\\/lib\\\/private\\\/AppFramework\\\/App.php(111): OC\\\\AppFramework\\\\Http\\\\Dispatcher->dispatch(Object(OCA\\\\Polls\\\\Controller\\\\PageController), 'insertPoll')\\n#10 \\\/var\\\/www\\\/html\\\/cloud\\\/lib\\\/private\\\/AppFramework\\\/Routing\\\/RouteActionHandler.php(47): OC\\\\AppFramework\\\\App::main('PageController', 'insertPoll', Object(OC\\\\AppFramework\\\\DependencyInjection\\\\DIContainer), Array)\\n#11 [internal function]: OC\\\\AppFramework\\\\Routing\\\\RouteActionHandler->__invoke(Array)\\n#12 \\\/var\\\/www\\\/html\\\/cloud\\\/lib\\\/private\\\/Route\\\/Router.php(293): call_user_func(Object(OC\\\\AppFramework\\\\Routing\\\\RouteActionHandler), Array)\\n#13 \\\/var\\\/www\\\/html\\\/cloud\\\/lib\\\/base.php(983): OC\\\\Route\\\\Router->match('\\\/apps\\\/polls\\\/ins...')\\n#14 \\\/var\\\/www\\\/html\\\/cloud\\\/index.php(48): OC::handleRequest()\\n#15 {main}\",\"File\":\"\\\/var\\\/www\\\/html\\\/cloud\\\/3rdparty\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/Driver\\\/AbstractMySQLDriver.php\",\"Line\":71}","level":3,"time":"2016-10-13T07:45:08+00:00","method":"POST","url":"\/cloud\/index.php\/apps\/polls\/insert","user":"user"}
```

**Comment by v1r0x** [comment]
Yes, the changes from the enh-group-display have not been merged. Will do it in the next days.

The database from master branch is missing a column in the polls table (`is_anonymous`). You could let the app add it by increasing the app version in `appinfo/info.xml`, but I think this is unnecessary, because the code change is already in the enc-group-display. I'll create a new PR after #80 has been merged.

**Comment by v1r0x** [comment]
I switched the milestone, since this is no critical bug and I think it is time to finally release 0.7.2.

**Comment by v1r0x** [comment]
0.7.2 has been released. Do you still have the wrong link?

**Comment by scroom** [comment]
Yes, still the same. If I change the link manually in the browser I can access the poll.

**Comment by v1r0x** [comment]
This is very weird. Neither on my test machine nor on my productive server I have this problem. Maybe there is another method I could use to generate the URL. I'll have a look at the API and create a PR/branch if I find something.

Do you have any other apps where (correct) links are present? Looking at their code could be helpful.

**Comment by scroom** [comment]
It is working with Files, calendar, contacts.

**Comment by v1r0x** [comment]
Couldn't find anything related to your issue. The other apps use the `linkToRouteAbsolute()` method as well. Maybe it's a "wrong" setting for your server's root dir or something like that? But still weird, that it works for a couple of apps, but not for another.
You could try to replace https://github.com/v1r0x/polls/blob/master/templates/goto.tmpl.php#L73 with
`$pollUrl = $urlGenerator->getAbsoluteURL($urlGenerator->linkToRoute('polls.page.goto_poll', ['hash' => $poll->getHash()]));`

**Comment by scroom** [comment]
I've changed as you mentioned but it didn't change the wrong URL. Maybe something about my overwrite.cli.url in the config.php of OC?

**Comment by v1r0x** [comment]
Could be the problem. Unfortunately, the URLGenerator only has the `linkToRoute()` and `linkToRouteAbsolute()` methods. I'll have a look at the API if there is another possibility to create URLs.

Do you get the correct URL if you set `overwrite.cli.url` to `https://example.com`?

**Comment by scroom** [comment]
The changing of the overwrite.cli.url had no effect.

**Comment by v1r0x** [comment]
Bummer...Maybe something in the apache/nginx/... config?

**Comment by scroom** [comment]
Hm, I have no idea. As everythin else is working as expected I don't think so (In fact I've done nothing special: Installed Apache from distribution channel (Ubuntu), just configured encryption, installed owncloud via the debs hosted on opensuse and tried as well to install Nextcloud following the official installation instructions. Both OC an NC worked fine. Just polls created these false links)

**Comment by v1r0x** [comment]
Unfortunately, I don't have an idea, maybe someone from nc can help? @BernhardPosselt ? :)

**Comment by BernhardPosselt** [comment]
The CLI URL setting on my instance does not contain the domain

**Comment by BernhardPosselt** [comment]
But that setting is only used when you run CLI commands. Are you sure the URL is not persisted somewhere (database/opcode cache)?

**Comment by BernhardPosselt** [comment]
You can clear the cache by restarting your webserver fyi

**Comment by scroom** [comment]
@BernhardPosselt I've restarted the webserver several times (its our test-server). Maybe it is something about a rewrite for SSL? But as mentioned before: the creation of links of core parts of NC is working as expected.

**Comment by michag86** [comment]
@scroom @v1r0x @BernhardPosselt can somebody test this?

**Comment by larseidam** [comment]
I have test it with sub folder and without
with subfolder you get:
**https://foo.bar.de/cloud//cloud/index.php/apps/polls/poll/UevZqkOsS0gzTILh**
without subfolder you get:
**https://foo.bar.de//index.php/apps/polls/poll/UevZqkOsS0gzTILh**
I'm running a Nextcloud 12.0.2 with Polls Version 0.7.3

**Comment by dartcafe** [comment]
I tried to reproduce the error. I set up an instance in a subfolder and I can't get the error. @scroom @larseidam: Do you have a base URL parameter set in the header of the html output?

**Comment by michag86** [comment]
@larseidam can you test this changes  #150 ?
@dartcafe the link is generated wrong at this position:
![grafik](https://user-images.githubusercontent.com/6691165/30113481-229c8f52-9315-11e7-9ce3-492217b67e04.png)

If you click on the name of the poll everything is fine.

**Comment by dartcafe** [comment]
Ah. Is see. I can confirm the error. Since it is the same URL, like the direkt link, it is OK to replace <?php p(OCP\Util::linkToAbsolute('', $pollUrl)); ?> with <?php p($pollUrl); ?>
